storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCacheTest.java

```diff
@@ -26,83 +26,136 @@
+                        segmentId, time.milliseconds(), Optional.empty(), state, brokerId1);
+                try {
+                    cache.updateRemoteLogSegmentMetadata(updatedMetadata);
+                    if (isInitialized) {
+                        fail("Should throw RemoteResourceNotFoundException when cache is initialized");
+                    }
+                } catch (RemoteResourceNotFoundException ex) {
+                    if (!isInitialized) {
+                        fail("Should not throw RemoteResourceNotFoundException when cache is not initialized");
+                    }
+                }
+            }
+        }
     }
 
-    private RemoteLogSegmentMetadata createSegmentUpdateWithState(RemoteLogMetadataCache cache,
-                                                                  Map<Integer, Long> segmentLeaderEpochs,
-                                                                  long startOffset,
-                                                                  long endOffset,
-                                                                  RemoteLogSegmentState state)
-            throws RemoteResourceNotFoundException {
-        RemoteLogSegmentId segmentId = new RemoteLogSegmentId(TP0, Uuid.randomUuid());
-        RemoteLogSegmentMetadata segmentMetadata = new RemoteLogSegmentMetadata(segmentId, startOffset, endOffset, -1L,
-                                                                                BROKER_ID_0, time.milliseconds(), SEG_SIZE, segmentLeaderEpochs);
+    @Test
+    public void testDropEventOnInvalidStateTransition() throws RemoteResourceNotFoundException {
+        cache.markInitialized();
+        int leaderEpoch = 5;
+        long offset = 10L;
+        RemoteLogSegmentId segmentId = new RemoteLogSegmentId(tpId0, Uuid.randomUuid());
+        RemoteLogSegmentMetadata segmentMetadata = new RemoteLogSegmentMetadata(segmentId, offset, 100L,
+                -1L, brokerId0, time.milliseconds(), segmentSize, Collections.singletonMap(leaderEpoch, offset));
         cache.addCopyInProgressSegment(segmentMetadata);
 
-        RemoteLogSegmentMetadataUpdate segMetadataUpdate = new RemoteLogSegmentMetadataUpdate(
-                segmentId,
-                time.milliseconds(),
-                Optional.empty(),
-                state, BROKER_ID_1);
-        cache.updateRemoteLogSegmentMetadata(segMetadataUpdate);
+        // invalid-transition-1. COPY_SEGMENT_STARTED -> DELETE_SEGMENT_FINISHED
+        RemoteLogSegmentMetadataUpdate updatedMetadata = new RemoteLogSegmentMetadataUpdate(segmentId, time.milliseconds(),
+                Optional.empty(), RemoteLogSegmentState.DELETE_SEGMENT_FINISHED, brokerId1);
+        updateAndVerifyCacheContents(updatedMetadata, RemoteLogSegmentState.COPY_SEGMENT_STARTED, leaderEpoch);
+
+        // valid-transition-2: COPY_SEGMENT_STARTED -> COPY_SEGMENT_FINISHED
```

Comment:
Nit: can you mark/split the invalid and valid transitions more clearly?

Reply:
To test all the variants of invalid transitions, we have to transition to some valid state in-between so incorporated all of them.